draft-private-access-tokens.md

```diff
@@ -173,14 +185,83 @@ Mediators advertise an Oblivious HTTP URI template {{!RFC6570}} for proxying
 protocol messages to Issuers. For example, one template
 for the Mediator might be https://mediator.net/relay-access-token-request.
 
-# Issuance
+# Protocol
+
+Private Access Tokens are single-use tokens cryptographically bound to
+policies. Redeemers request tokens from Clients, who then engage with
+Mediators and Issuers to private compute policy-compliant tokens and
+reveal them to the Redeemer. Example policies and use cases that system
+addresses are described in {{examples}}.
+
+The rest of this section describes this interactive protocol in terms of
+the redemption request ({{redemption-request}}) and corresponding token
+issuance flow ({{issuance}}).
+
+## Redemption Request
+
+The Client is assumed to have the policy verification key before redeeming
+a Private Access Token. See {{access-token-keys}} for details on how this
+can be obtained.
+
+Token redemption is an interactive protocol. Redeemers challenge Clients to
+present a unique, single-use token. Redeemers present this challenge to Clients
+with the following context:
+
+~~~
+struct {
+   uint8 policy_window;
+   opaque redeemer_name<1..2^16-1>;
+   opaque issuer_name<1..2^16-1>;
+   opaque redemption_nonce[32];
+   uint64 timestamp;
+} RedemptionContext;
+~~~
+
+policy_window:
+: The policy window value for the redemption token (ISSUER_POLICY_WINDOW).
+
+redeemer_name:
+: Name of the Redeemer (ORIGIN_NAME).
+
+issuer_name:
+: Name of the Issuer (ISSUER_NAME).
+
+redemption_nonce:
+: A fresh 32-byte nonce generated for each redemption request.
+
+timestamp:
+: The current UNIX epoch in seconds.
+
+This context is sent to Clients in an "WWW-Authenticate" header as follows:
+
+~~~
+WWW-Authenticate: PrivacyToken realm="<context>"
+~~~
+
+Upon receipt, Clients use the context in the Issuance protocol, described in {{issuance}}.
+The output of that protocol is a PrivacyToken bound to the redemption context.
+Clients then present this token to Redeemers using the Authorization header as follows:
 
-Issuance assumes the Client has the following information:
+~~~
+Authorization: PrivacyToken t=abc
+~~~
+
+Where the token is a serialized Private Access Token corresponding to the given Redeemer
+policy, and the PrivacyToken message is SHA256(context). Redeemers verify the token using
+the corresponding policy verification key from the Issuer.
```

Comment:
Presumably it would be bound to the session, or clients could echo back the context when presenting a token?

Reply:
With the interactive mode, I think it would need to be tied to the HTTP session, right?

Reply:
Yeah, I don't think the client could be echoing back, otherwise it would lie about the previous nonce (unless you have some sort of signed bit there).